### PR TITLE
Local context parser

### PR DESCRIPTION
**Pull request by tronxd**
support of local variable evaluations, major refactor of context parsers and record class

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: checkov/terraform/context_parsers/base_parser.py [hunk 1]
```diff
@@ -68,43 +73,41 @@ def _compute_definition_end_line(self, start_line_num):
...
-            if not self.context.get(entity_type):
-                self.context[entity_type] = {}
-            if not self.context.get(entity_type).get(entity_name):
-                self.context[entity_type][entity_name] = {}
+        for i, entity_block in enumerate(definition_blocks):
+            entity_context_path = self.get_entity_context_path(entity_block)
             for line_num, line in parsed_file_lines:
                 line_tokens = [x.replace('"', "") for x in line.split()]
-                if all(x in line_tokens for x in [self.get_block_type(), entity_type, entity_name]):
+                if self._is_block_signature(line_tokens, entity_context_path):
```
Comment (schosterbarak) [inline checkov/terraform/context_parsers/base_parser.py:106]: is that returning a boolean value? name hints it is when it does not.

I know it's not part of the original scope, but can you please add class comments and methods comments to the common classes you have modified?

File: checkov/terraform/context_parsers/parsers/locals_context_parser.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def _collect_local_values(self, local_block):
+        for local_name, local_value in local_block.items():
+            local_value = local_value[0]
+            if type(local_value) in (int, float, bool, str):
+                dpath.new(self.context, ['assignments', local_name], local_value)
+
+    def get_block_type(self):
+        return self.definition_type
+
+    def get_entity_context_path(self, entity_block):
```
Comment (schosterbarak) [inline checkov/terraform/context_parsers/parsers/locals_context_parser.py:19]: what is this method used for?

File: checkov/terraform/context_parsers/parsers/resource_context_parser.py [hunk 3]
```diff
@@ -3,11 +3,18 @@
 
 class ResourceContextParser(BaseContextParser):
     def __init__(self):
-        definition_type = 'resource'
+        definition_type = "resource"
         super().__init__(definition_type=definition_type)
 
-    def get_block_type(self):
-        return self.definition_type
+    def _is_block_signature(self, line_tokens, entity_context_path):
```
Comment (schosterbarak) [inline checkov/terraform/context_parsers/parsers/resource_context_parser.py:9]: "_" prefix is for private methods. but you call this method outside of this class - maybe that should be changed?

**Comment by schosterbarak** [comment]
Overall very good. let's talk about some minor comments today :)

File: checkov/terraform/context_parsers/base_parser.py (hunk 1 above)
Comment (tronxd) [reply checkov/terraform/context_parsers/base_parser.py:106]: @schosterbarak It's actually returns a boolean value
docs are on the way anyway :)

File: checkov/terraform/context_parsers/parsers/locals_context_parser.py (hunk 2 above)
Comment (tronxd) [reply checkov/terraform/context_parsers/parsers/locals_context_parser.py:19]: @schosterbarak given an entity (may be a `resource`, `data`, `provider` etc.), return the entity's path in the context parser object. It is an overridden method in each context parser that implements the base context parser.

File: checkov/terraform/context_parsers/parsers/resource_context_parser.py (hunk 3 above)
Comment (tronxd) [reply checkov/terraform/context_parsers/parsers/resource_context_parser.py:9]: @schosterbarak Thanks, it's a code leftover. the implementation is in `base_context_parser` and usage is only class-internal.

File: checkov/terraform/context_parsers/parsers/locals_context_parser.py (hunk 2 above)
Comment (schosterbarak) [reply checkov/terraform/context_parsers/parsers/locals_context_parser.py:19]: great. maybe we should add that in a pydoc on that method? :)

**Pull request merged**